{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 92,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Defaulting to user installation because normal site-packages is not writeable\n",
            "Requirement already satisfied: biopython in /Users/ed282828/Library/Python/3.9/lib/python/site-packages (1.83)\n",
            "Requirement already satisfied: numpy in /Users/ed282828/Library/Python/3.9/lib/python/site-packages (from biopython) (1.26.4)\n",
            "Note: you may need to restart the kernel to use updated packages.\n"
          ]
        }
      ],
      "source": [
        "pip install biopython"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 93,
      "metadata": {},
      "outputs": [],
      "source": [
        "from Bio import Entrez\n",
        "Entrez.email = 'learnbiopython@gmail.com'"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 98,
      "metadata": {},
      "outputs": [],
      "source": [
        "term1 = [\n",
        "    'Bos gaurus', 'Antelope cervicapra', 'Gazella bennettii', 'Boselaphus tragocamelus',\n",
        "    'Canis lupus', 'Elephas maximus',\n",
        "    'Cervus canadensis', 'Pavo cristatus', 'Grus leucogeranus', 'Vulpes vulpes',\n",
        "    'Rhinoceros unicornis', 'Panthera Tigris', 'Crocodylus palustris', 'Gavialis gangeticus',\n",
        "    'Equus caballus', 'Equus quagga', 'Babalus bubalis', 'Sus scrofa', 'Camelus dromedaries',\n",
        "    'Giraffa camelopardalis ', 'Hemidactylus flaviviridis', 'Hippopotamus amphibius',\n",
        "    'Macaca mulatta', 'Canis lupus', 'Felis domesticus', 'Acinonyx jubatus', 'Rattus rattus',\n",
        "    'Mus musculus', 'Oryctolagus cuniculus', 'Bubo virginianus', 'Passer domesticus',\n",
        "    'Corvus splendens', 'Acridotheres tristis', 'Psittacula eupatria', 'Molpastes cafer',\n",
        "    'Eudynamis scolopaccus', 'Columba livia', 'Naja naja', 'Ophiophagus hannah',\n",
        "    'Hydrophiinae ', 'Python molurus', 'Ptyas mucosa'\n",
        "]\n",
        "\n",
        "term2 = [\n",
        "    'Bos gaurus', 'Gazella bennettii', 'Boselaphus tragocamelus',\n",
        "    'Canis lupus', 'Elephas maximus',  \n",
        "    'Cervus canadensis', 'Pavo cristatus', 'Grus leucogeranus', 'Vulpes vulpes',\n",
        "    'Crocodylus palustris', 'Gavialis gangeticus',\n",
        "    'Equus caballus', 'Equus quagga', 'Sus scrofa', 'Camelus dromedaries',\n",
        "    'Canis lupus', 'Felis domesticus','Rattus rattus',\n",
        "    'Mus musculus', 'Oryctolagus cuniculus'\n",
        "]\n",
        "term3 = ['Tursiops truncatus']\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 99,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "eFcMM1ofABTx",
        "outputId": "bf2116d5-b7e1-4822-a5a8-51f4630b783f"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "__________________________________________________________________________________\n",
            " |          Given Names           |     Binomial Nomenclature      |   Status   |\n",
            "__________________________________________________________________________________\n",
            " |           Bos gaurus           |           Bos gaurus           |    True    |\n",
            " |      Antelope cervicapra       |      Antelope cervicapra       |    True    |\n",
            " |       Gazella bennettii        |       Gazella bennettii        |    True    |\n",
            " |    Boselaphus tragocamelus     |    Boselaphus tragocamelus     |    True    |\n",
            " |          Canis lupus           |          Canis lupus           |    True    |\n",
            " |        Elephas maximus         |        Elephas maximus         |    True    |\n",
            " |       Cervus canadensis        |       Cervus canadensis        |    True    |\n",
            " |         Pavo cristatus         |         Pavo cristatus         |    True    |\n",
            " |       Grus leucogeranus        |       Grus leucogeranus        |    True    |\n",
            " |         Vulpes vulpes          |         Vulpes vulpes          |    True    |\n",
            " |      Rhinoceros unicornis      |      Rhinoceros unicornis      |    True    |\n",
            " |        Panthera Tigris         |        Panthera tigris         |   False    |\n",
            " |      Crocodylus palustris      |      Crocodylus palustris      |    True    |\n",
            " |      Gavialis gangeticus       |      Gavialis gangeticus       |    True    |\n",
            " |         Equus caballus         |         Equus caballus         |    True    |\n",
            " |          Equus quagga          |          Equus quagga          |    True    |\n",
            " |        Babalus bubalis         |        Bubalus bubalis         |   False    |\n",
            " |           Sus scrofa           |           Sus scrofa           |    True    |\n",
            " |      Camelus dromedaries       |      Camelus dromedaries       |    True    |\n",
            " |    Giraffa camelopardalis      |     Giraffa camelopardalis     |   False    |\n",
            " |   Hemidactylus flaviviridis    |  N hemidactylus flaviviridis   |   False    |\n",
            " |     Hippopotamus amphibius     |     Hippopotamus amphibius     |    True    |\n",
            " |         Macaca mulatta         |         Macaca mulatta         |    True    |\n",
            " |          Canis lupus           |          Canis lupus           |    True    |\n",
            " |        Felis domesticus        |        Felis domesticus        |    True    |\n",
            " |        Acinonyx jubatus        |        Acinonyx jubatus        |    True    |\n",
            " |         Rattus rattus          |         Rattus rattus          |    True    |\n",
            " |          Mus musculus          |          Mus musculus          |    True    |\n",
            " |     Oryctolagus cuniculus      |     Oryctolagus cuniculus      |    True    |\n",
            " |        Bubo virginianus        |        Bubo virginianus        |    True    |\n",
            " |       Passer domesticus        |       Pass er domesticus       |   False    |\n",
            " |        Corvus splendens        |        Corvus splendens        |    True    |\n",
            " |      Acridotheres tristis      |      Acridotheres tristis      |    True    |\n",
            " |      Psittacula eupatria       |      Psittacula eupatria       |    True    |\n",
            " |        Molpastes cafer         |        Omlpasses cafer         |   False    |\n",
            " |     Eudynamis scolopaccus      |     Eudynamys scolopaceus      |   False    |\n",
            " |         Columba livia          |         Columba livia          |    True    |\n",
            " |           Naja naja            |           Naja naja            |    True    |\n",
            " |       Ophiophagus hannah       |       Ophiophagus hannah       |    True    |\n",
            " |         Hydrophiinae           |          Hydrophiidae          |   False    |\n",
            " |         Python molurus         |         Python molurus         |    True    |\n",
            " |          Ptyas mucosa          |          Ptyas mucosa          |    True    |\n",
            "__________________________________________________________________________________\n"
          ]
        }
      ],
      "source": [
        "\n",
        "def checkNames(term):\n",
        "    handle = Entrez.espell(db='pmc', term=term)\n",
        "    record = Entrez.read(handle)\n",
        "    handle.close()\n",
        "    correctTerm = record.get('CorrectedQuery', term)\n",
        "    words = correctTerm.split()\n",
        "    correctWord = [word.capitalize() if i == 0 else word.lower() for i, word in enumerate(words)]\n",
        "    return \" \".join(correctWord)\n",
        "\n",
        "print(\"_\" * 82)\n",
        "print(f\" | {'Given Names':^30} | {'Binomial Nomenclature':^30} | {'Status':^10} |\")\n",
        "print(\"_\" * 82)\n",
        "\n",
        "for name in term1:\n",
        "    correctTerm = checkNames(name)\n",
        "    status = \"True\" if name == correctTerm else \"False\"\n",
        "    correctDisplay = correctTerm if status == \"False\" else name\n",
        "    print(f\" | {name:^30} | {correctDisplay:^30} | {status:^10} |\")\n",
        "print(\"_\" * 82)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 100,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "oAmCQBN0AJ1W",
        "outputId": "42c65cd1-83e7-4843-e1b1-6bf7d286f3e8"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "______________________________________________________________________________________________________________\n",
            " |           Nama Ilmiah           |                                Nama Umum                                |\n",
            "______________________________________________________________________________________________________________\n",
            " |           Bos gaurus            |                                seladang                                 |\n",
            " |        Gazella bennettii        |                             Indian gazelle                              |\n",
            " |     Boselaphus tragocamelus     |                           blue bull, nilghai                            |\n",
            " |           Canis lupus           |                                grey wolf                                |\n",
            " |         Elephas maximus         |                     Asian elephant, Indian elephant                     |\n",
            " |        Cervus canadensis        |                               elk, wapiti                               |\n",
            " |         Pavo cristatus          |                              blue peafowl                               |\n",
            " |        Grus leucogeranus        |                               white crane                               |\n",
            " |          Vulpes vulpes          |                               silver fox                                |\n",
            " |      Crocodylus palustris       |           Indian crocodile, Indus crocodile, mugger crocodile           |\n",
            " |       Gavialis gangeticus       |                              true gharial                               |\n",
            " |         Equus caballus          |                         domestic horse, equine                          |\n",
            " |          Equus quagga           |                              plains zebra                               |\n",
            " |           Sus scrofa            |                         pigs, swine, wild boar                          |\n",
            " |       Camelus dromedaries       |    camel, dromedaries, dromedary, dromedary camel, one-humped camel     |\n",
            " |           Canis lupus           |                                grey wolf                                |\n",
            " |        Felis domesticus         |                                cat, cats                                |\n",
            " |          Rattus rattus          |                           house rat, roof rat                           |\n",
            " |          Mus musculus           |                                  mouse                                  |\n",
            " |      Oryctolagus cuniculus      |    European rabbit, Japanese white rabbit, domestic rabbit, rabbits     |\n"
          ]
        }
      ],
      "source": [
        "\n",
        "def commonNames(terms):\n",
        "\n",
        "    handle = Entrez.esearch(db='taxonomy', term=terms)\n",
        "    record = Entrez.read(handle)\n",
        "    handle.close()\n",
        "\n",
        "    if record['IdList']:\n",
        "        tax_id = record['IdList'][0]\n",
        "        handle = Entrez.efetch(db='taxonomy', id=tax_id)\n",
        "        tax_record = Entrez.read(handle)\n",
        "        handle.close()\n",
        "\n",
        "        if 'CommonName' in tax_record[0]['OtherNames']:\n",
        "            common_names = tax_record[0]['OtherNames']['CommonName']\n",
        "            if isinstance(common_names, list):\n",
        "                return \", \".join(common_names)\n",
        "            else:\n",
        "                return common_names\n",
        "    else:\n",
        "        return \"N/A\"\n",
        "\n",
        "print(\"_\" * 110)\n",
        "print(f\" | {'Nama Ilmiah':^31} | {'Nama Umum':^71} |\")\n",
        "print(\"_\" * 110)\n",
        "\n",
        "for name in term2:\n",
        "    common_name = commonNames(name)\n",
        "    print(f\" | {name:^31} | {common_name:^71} |\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 101,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Scientific Name: Tursiops truncatus\n",
            "Taxonomy Rank: species\n",
            "\n",
            "Lineage:\n",
            "_____________________________________\n",
            " | Level |          Taxon           |\n",
            "_____________________________________\n",
            " |   1   |    cellular organisms    |\n",
            " |   2   |         Eukaryota        |\n",
            " |   3   |       Opisthokonta       |\n",
            " |   4   |          Metazoa         |\n",
            " |   5   |         Eumetazoa        |\n",
            " |   6   |         Bilateria        |\n",
            " |   7   |       Deuterostomia      |\n",
            " |   8   |         Chordata         |\n",
            " |   9   |         Craniata         |\n",
            " |  10   |        Vertebrata        |\n",
            " |  11   |       Gnathostomata      |\n",
            " |  12   |        Teleostomi        |\n",
            " |  13   |       Euteleostomi       |\n",
            " |  14   |       Sarcopterygii      |\n",
            " |  15   |   Dipnotetrapodomorpha   |\n",
            " |  16   |         Tetrapoda        |\n",
            " |  17   |          Amniota         |\n",
            " |  18   |         Mammalia         |\n",
            " |  19   |          Theria          |\n",
            " |  20   |         Eutheria         |\n",
            " |  21   |       Boreoeutheria      |\n",
            " |  22   |      Laurasiatheria      |\n",
            " |  23   |       Artiodactyla       |\n",
            " |  24   |       Whippomorpha       |\n",
            " |  25   |          Cetacea         |\n",
            " |  26   |        Odontoceti        |\n",
            " |  27   |        Delphinidae       |\n",
            " |  28   |         Tursiops         |\n",
            "_____________________________________\n",
            "\n",
            "Related Genomic Publications:\n",
            "1. Title: Chromosome-level genome provides insights into environmental adaptability and innate immunity in the common dolphin (delphinus delphis).\n",
            "   Full Journal Name: BMC genomics\n",
            "   DOI: 10.1186/s12864-024-10268-4\n",
            "\n",
            "2. Title: First description of Rhodococcus equi infection in common bottlenose dolphin (Tursiops truncatus).\n",
            "   Full Journal Name: Veterinary research communications\n",
            "   DOI: 10.1007/s11259-024-10352-w\n",
            "\n",
            "3. Title: Genomic Divergence and the Evolution of Ecotypes in Bottlenose Dolphins (Genus Tursiops).\n",
            "   Full Journal Name: Genome biology and evolution\n",
            "   DOI: 10.1093/gbe/evad199\n",
            "\n",
            "4. Title: Recent selection created distinctive variability patterns on MHC class II loci in three dolphin species from the Mediterranean Sea.\n",
            "   Full Journal Name: Developmental and comparative immunology\n",
            "   DOI: 10.1016/j.dci.2023.105079\n",
            "\n",
            "5. Title: Proteomics as a Metrological Tool to Evaluate Genome Annotation Accuracy Following De Novo Genome Assembly: A Case Study Using the Atlantic Bottlenose Dolphin (Tursiops truncatus).\n",
            "   Full Journal Name: Genes\n",
            "   DOI: 10.3390/genes14091696\n",
            "\n",
            "Genome Database Size: 149731 entries\n",
            "\n",
            "GenBank Information:\n",
            ">NM_001289846.2 Tursiops truncatus homeobox C6 (HOXC6), transcript variant 2, mRNA\n",
            "AACTTTTTATTGTGGTTTGTCCGTTCGGAGCGCTCCACAGAACAGTCCTCCCTGTAAGAGCCTAACCATT\n",
            "GCCAGGGAAACCTGCGCTGGGCGCTCCCTTCATTACCAGTATTTTTTTAAATTAATCTGATTAATAATTA\n",
            "TTTTCTCCCCTCATTTAATTTCTTTCCTCCCAGATGGAGTTGCTGGAAGCTGGAAGCACCTGGAGAGGGG\n",
            "GGTGGGGGGAGAGGCAGGAGTAGAGGACACCCCTCTCTCCCTTCCCGACCCTCTAGCCCCCCAAGGGGCA\n",
            "GGAGGAATGCGGGAGCAGTTGAGATAGGGAGCTGCAGCTGCCTCCGCCCCTCCCCTCTCCCAGGCTCTCA\n",
            "CTCCTGCCTCCTTCTCGCAACTCTGCTTCAAATTTATTTGGCTTTTTGATTATTGGGACTACTCTTTTTT\n",
            "ATTTTTAAATTTATATAAAGTATATATATGTATGTGGAGCTGAAGCAGTCTCGGCAGCGGCACAGAATGA\n",
            "GGGAAGACGAGAAAGAGAGAGAGTGGGAGAGAGAGGCAGAGAGAGAGAGAGGGAGAGGGAGAGTGACAGC\n",
            "AGCGCCCGGACGTCCTCCCCAACGTCGCCCTCAATTCCACCGCCTATGATCCAGTGAGGCATTTCTCGAC\n",
            "CTATGGAGCGGCCGTTGCCCAGAACCGGATCTACTCGACTCCCTTTTATTCGCCACAGGAGAATGTCGTG\n",
            "TTCAGTTCCAGCCGAGGGCCGTATGACTATGGATCTAATTCCTTTTACCAGGAGAAAGACATGCTCTCAA\n",
            "ACTGCAGACAAAACACCTTAGGACATAACACACAGACCTCAATCGCTCAGGATTTTAGTTCTGAGCAGGG\n",
            "CAGGACTGCGCCCCAGGACCAGAAAGCCAGTATCCAGATTTACCCCTGGATGCAGCGAATGAATTCGCAC\n",
            "AGTGGGGTCGGCTATGGAGCGGACCGGAGGCGCGGTCGCCAGATCTACTCGCGGTACCAGACCCTGGAAC\n",
            "TGGAGAAGGAATTTCACTTCAACCGCTACCTAACGCGGCGCCGGCGCATCGAGATCGCCAACGCGCTCTG\n",
            "CCTGACCGAGCGACAGATCAAAATCTGGTTCCAGAACCGCCGGATGAAGTGGAAAAAAGAGTCTAATCTC\n",
            "ACGTCCACGCTCTCAGGGGGCGGCGGAGGGGCCGCGGTCGACAGCCTGGGCGGAAAAGAGGAAAAGCGGG\n",
            "AAGAGACAGAAGAGGAGAAGCAGAAAGAGTGACCAGGACTGTCCCTGCCGCCCCTCTCTCTCCTGCTCCC\n",
            "TCGCTCCCCGCAACTCCCTCCTAATCACACACTCCGTATTTATCACTGGCACAATTGATGTTTGGACTTC\n",
            "CTGAGACAAAATTAGGGAGTCAAATATGGACCTGAAAAGTCAGCTCTGGACCCCCTCCCTCACCGCACAG\n",
            "ACTCACAAGGCGCCTCCTCCTCTCCTGCTCCCTCGCTAGCTCGTTCTCGGCTCGTCTGCAGGCCCCTTCC\n",
            "CTCACCCAGGCCTTGGGGCTCGTGCCTGAACCTCGCTTCAGACTCCACAGATCTCCCTCCAAACGAGGAC\n",
            "TTAGCCTCCCACCCTCTCGGCGCCCCTCACCCCCGCCCCCGCGCAGAGAGCCGGCTCCCGGGCCCGGGGG\n",
            "CCTCCGCTCCCAGGCCTCAGGGCCCGGCGTGGCAGCCGGGGAGGGCGGGAGGCCCAAGTAGGGCGCGTGG\n",
            "TGATCCCACAGCAACCCCCAGGGCCTGCCCGCAGCCCCCGCCCGGCCCCTCTGCCCCAGCAAATGCCCAG\n",
            "CCCAGGCGAATTGTATTTAAAGAAGCCTGGGGATCATTATGGCATATTACAAACTGTGACCATTTCTGTG\n",
            "TGAATATTTTTAGCTGTATTTGTGGTCTCTGTATTTATATTTATGTTTAGCACCGTCCGTGTTCCAATCC\n",
            "CATCTTAAAAAGGAAAAAAAAAAAGAGGGAAAAGTACAAAAAGAGAGAAAAAATAGTGAATGACGTTTGT\n",
            "TTAGCCGGTAGGAGAAAAATAATAAATTTTAAAAATCCCGTCGTGTTACCCTCCTGTATAAATCCGACCT\n",
            "CTGGATCCGTTCTCGAATATTTAATAAAACTGATATTATTTTTAAAAGTTTA\n",
            "\n",
            "\n",
            "<class 'Bio.Entrez.Parser.DictionaryElement'>\n",
            "dict_keys(['Count', 'RetMax', 'RetStart', 'IdList', 'TranslationSet', 'TranslationStack', 'QueryTranslation'])\n",
            "Count : 132374\n",
            "RetMax : 10\n",
            "RetStart : 0\n",
            "IdList : ['11146340', '11146323', '11146197', '11145504', '11144905', '11144893', '11144879', '11144871', '11144625', '11144543']\n",
            "TranslationSet : [{'From': 'Felis domesticus', 'To': '\"cats\"[MeSH Terms] OR \"cats\"[All Fields] OR (\"felis\"[All Fields] AND \"domesticus\"[All Fields]) OR \"felis domesticus\"[All Fields]'}]\n",
            "TranslationStack : [{'Term': '\"cats\"[MeSH Terms]', 'Field': 'MeSH Terms', 'Count': '14980', 'Explode': 'Y'}, {'Term': '\"cats\"[All Fields]', 'Field': 'All Fields', 'Count': '131364', 'Explode': 'N'}, 'OR', {'Term': '\"felis\"[All Fields]', 'Field': 'All Fields', 'Count': '16352', 'Explode': 'N'}, {'Term': '\"domesticus\"[All Fields]', 'Field': 'All Fields', 'Count': '20373', 'Explode': 'N'}, 'AND', 'GROUP', 'OR', {'Term': '\"felis domesticus\"[All Fields]', 'Field': 'All Fields', 'Count': '4958', 'Explode': 'N'}, 'OR', 'GROUP']\n",
            "QueryTranslation : \"cats\"[MeSH Terms] OR \"cats\"[All Fields] OR (\"felis\"[All Fields] AND \"domesticus\"[All Fields]) OR \"felis domesticus\"[All Fields]\n"
          ]
        }
      ],
      "source": [
        "def fetch_taxonomy_info(terms):\n",
        "    try:\n",
        "        handle = Entrez.esearch(db='taxonomy', term=terms)\n",
        "        record = Entrez.read(handle)\n",
        "        handle.close()\n",
        "        if record['IdList']:\n",
        "            tax_id = record['IdList'][0]\n",
        "            handle = Entrez.efetch(db='taxonomy', id=tax_id)\n",
        "            tax_record = Entrez.read(handle)\n",
        "            handle.close()\n",
        "            return tax_record[0]\n",
        "    except Exception as e:\n",
        "        print(f\"Error fetching taxonomy information for {terms}: {e}\")\n",
        "        return None\n",
        "\n",
        "def fetch_genomic_publications(terms):\n",
        "    try:\n",
        "        record = Entrez.read(Entrez.esearch(db='pubmed', term=f\"{terms} genome\", retmax=5))\n",
        "        if record['IdList']:\n",
        "            publications = []\n",
        "            for pub_id in record['IdList']:\n",
        "                pub_summary = Entrez.read(Entrez.esummary(db='pubmed', id=pub_id))\n",
        "                publications.append({\n",
        "                    'Title': pub_summary[0]['Title'],\n",
        "                    'FullJournalName': pub_summary[0]['FullJournalName'],\n",
        "                    'DOI': pub_summary[0].get('DOI', 'N/A')\n",
        "                })\n",
        "            return publications\n",
        "    except Exception as e:\n",
        "        print(f\"Error fetching genomic publications for {terms}: {e}\")\n",
        "        return []\n",
        "\n",
        "def fetch_genome_database_size(terms):\n",
        "    try:\n",
        "        record = Entrez.read(Entrez.esearch(db='nucleotide', term=f\"{terms} genome\", retmax=1))\n",
        "        return record['Count']\n",
        "    except Exception as e:\n",
        "        print(f\"Error fetching genome database size for {terms}: {e}\")\n",
        "        return '0'\n",
        "\n",
        "def fetch_genbank_info(terms):\n",
        "    try:\n",
        "        record = Entrez.read(Entrez.esearch(db='nucleotide', term=f'{terms}[Organism] AND refseq[filter]', retmax=1, idtype='acc'))\n",
        "        if record['IdList']:\n",
        "            ID = record['IdList'][0]\n",
        "            if 'NM_' in ID:\n",
        "                fetch = Entrez.efetch(db='nucleotide', id=ID, rettype='fasta', retmode='text')\n",
        "                read_fetch = fetch.read()\n",
        "                return read_fetch\n",
        "    except Exception as e:\n",
        "        print(f\"Error fetching GenBank information for {terms}: {e}\")\n",
        "        return None\n",
        "\n",
        "def process_termss(termss):\n",
        "    for name in termss:\n",
        "        tax_info = fetch_taxonomy_info(name)\n",
        "        if tax_info:\n",
        "            print(f\"Scientific Name: {tax_info['ScientificName']}\")\n",
        "            print(f\"Taxonomy Rank: {tax_info['Rank']}\")\n",
        "\n",
        "            lineage_list = tax_info['Lineage'].split(\"; \")\n",
        "            print(\"\\nLineage:\")\n",
        "            print(\"_\" * 37)\n",
        "            print(f\" | {'Level':^5} | {'Taxon':^24} |\")\n",
        "            print(\"_\" * 37)\n",
        "            for level, taxon in enumerate(lineage_list, 1):\n",
        "                print(f\" | {level:^5} | {taxon:^25}|\")\n",
        "            print(\"_\" * 37)\n",
        "\n",
        "            print(\"\\nRelated Genomic Publications:\")\n",
        "            publications = fetch_genomic_publications(name)\n",
        "            for idx, pub in enumerate(publications, 1):\n",
        "                print(f\"{idx}. Title: {pub['Title']}\")\n",
        "                print(f\"   Full Journal Name: {pub['FullJournalName']}\")\n",
        "                print(f\"   DOI: {pub['DOI']}\\n\")\n",
        "\n",
        "            genome_size = fetch_genome_database_size(name)\n",
        "            print(f\"Genome Database Size: {genome_size} entries\")\n",
        "\n",
        "            genbank_info = fetch_genbank_info(name)\n",
        "            if genbank_info:\n",
        "                print(\"\\nGenBank Information:\")\n",
        "                print(genbank_info)\n",
        "        else:\n",
        "            print(f\"No taxonomy information found for {name}.\")\n",
        "\n",
        "# Example list of scientific names to process\n",
        "process_termss(term3)\n",
        "\n",
        "# Fetch records from PMC\n",
        "try:\n",
        "    record = Entrez.read(Entrez.esearch(db='pmc', term='Felis domesticus', retmax=10))\n",
        "    print(type(record))\n",
        "    print(record.keys())\n",
        "    for key in record.keys():\n",
        "        print(key, ':', record[key])\n",
        "except Exception as e:\n",
        "    print(f\"Error fetching PMC records: {e}\")\n"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.9.6"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
